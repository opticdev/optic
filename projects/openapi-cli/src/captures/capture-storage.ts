import Path from 'path';
import * as fs from 'fs-extra';
import { InputErrors } from '../commands/reporters/feedback';
import os from 'os';
const tmpDirectory = os.tmpdir();
import crypto from 'crypto';

export async function captureStorage(
  filePath: string
): Promise<[boolean, string]> {
  const resolvedFilepath = Path.resolve(filePath);
  const openApiExists: boolean = await fs.pathExists(resolvedFilepath);

  const specPathHash = crypto
    .createHash('md5')
    .update(resolvedFilepath)
    .digest('hex');

  const trafficDirectory = Path.join(
    tmpDirectory,
    'optic',
    'captures',
    specPathHash
  );

  if (openApiExists) await fs.ensureDir(trafficDirectory);

  return [openApiExists, trafficDirectory];
}
