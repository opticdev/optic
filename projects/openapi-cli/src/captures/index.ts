import { CapturedBody } from './body';
import { CapturedBodies } from './streams/captured-bodies';
import { OpenAPIV3 } from '../specs';

export { CapturedBody, CapturedBodies };
export { CapturedInteractions } from './streams/captured-interactions';

export interface CapturedInteraction {
  request: {
    host: string;
    method: OpenAPIV3.HttpMethods;
    path: string;
    body: CapturedBody | null;
    // TODO: add support for headers and query params
  };
  response: {
    statusCode: string;
    body: CapturedBody | null;
    // TODO: add support headers
  };
}

export type CapturedRequest = CapturedInteraction['request'];
export type CapturedResponse = CapturedInteraction['response'];
