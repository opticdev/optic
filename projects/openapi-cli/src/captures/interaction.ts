import { CapturedBody } from './body';
import { OpenAPIV3 } from '../specs';
import { HttpArchive } from './streams/sources/har';
import { ProxySource } from './streams/sources/proxy';
import { URL } from 'url';
import { HttpMethods, Operation } from '../operations';
import invariant from 'ts-invariant';
import { Buffer } from 'buffer';

export interface CapturedInteraction {
  request: {
    host: string;
    method: OpenAPIV3.HttpMethods;
    path: string;
    body: CapturedBody | null;
    // TODO: add support for headers and query params
  };
  response: {
    statusCode: string;
    body: CapturedBody | null;
    // TODO: add support headers
  };
}
export class CapturedInteraction {
  static fromHarEntry(entry: HttpArchive.Entry): CapturedInteraction {
    const url = new URL(entry.request.url);

    const method = HttpMethods[entry.request.method];
    invariant(
      Operation.isHttpMethod(method),
      `expect HAR entry to have a valid request method`
    );

    let requestBody: CapturedBody | null = null;
    let responseBody: CapturedBody | null = null;

    const requestPostData = entry.request.postData;
    if (
      requestPostData &&
      (!requestPostData.encoding || Buffer.isEncoding(requestPostData.encoding))
    ) {
      let buffer = Buffer.from(
        requestPostData.text,
        requestPostData.encoding as BufferEncoding | undefined
      );
      requestBody = CapturedBody.from(
        buffer,
        requestPostData.mimeType,
        entry.request.bodySize
      );
    }

    const responseContent = entry.response.content;
    if (
      responseContent.text &&
      (!responseContent.encoding || Buffer.isEncoding(responseContent.encoding))
    ) {
      let buffer = Buffer.from(
        responseContent.text,
        responseContent.encoding as BufferEncoding | undefined
      );
      responseBody = CapturedBody.from(
        buffer,
        responseContent.mimeType,
        responseContent.size
      );
    }

    return {
      request: {
        host: url.hostname,
        method,
        path: url.pathname,
        body: requestBody,
      },
      response: {
        statusCode: '' + entry.response.status,
        body: responseBody,
      },
    };
  }

  static fromProxyInteraction(
    proxyInteraction: ProxySource.Interaction
  ): CapturedInteraction {
    const url = new URL(proxyInteraction.request.url);

    const method = HttpMethods[proxyInteraction.request.method];
    invariant(
      Operation.isHttpMethod(method),
      `expect proxy interaction to have a valid request method`
    );

    let requestBody: CapturedBody | null = null;
    let responseBody: CapturedBody | null = null;

    const requestBodyBuffer = proxyInteraction.request.body.buffer;
    if (requestBodyBuffer.length > 0) {
      let contentType = proxyInteraction.request.headers['content-type'];
      let contentLength = proxyInteraction.request.headers['content-length'];

      requestBody = CapturedBody.from(
        requestBodyBuffer,
        contentType || null,
        contentLength ? parseInt(contentLength, 10) : 0
      );
    }

    const responseBodyBuffer = proxyInteraction.response.body.buffer;
    if (responseBodyBuffer.length > 0) {
      let contentType = proxyInteraction.response.headers['content-type'];
      let contentLength = proxyInteraction.response.headers['content-length'];

      responseBody = CapturedBody.from(
        responseBodyBuffer,
        contentType || null,
        contentLength ? parseInt(contentLength, 10) : 0
      );
    }

    return {
      request: {
        host: url.hostname,
        method,
        path: url.pathname,
        body: requestBody,
      },
      response: {
        statusCode: '' + proxyInteraction.response.statusCode,
        body: responseBody,
      },
    };
  }
}

export type CapturedRequest = CapturedInteraction['request'];
export type CapturedResponse = CapturedInteraction['response'];
