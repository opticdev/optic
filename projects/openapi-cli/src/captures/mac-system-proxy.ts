export function parseMacNetworkLine(line: string): {
  enabled: boolean;
  host: string;
  port: string;
  authenticated: boolean;
} {
  const match = line.match(
    /Enabled: (Yes|No)\nServer: ([:/\da-zA-Z.\-_]*)\nPort: (\d+)\nAuthenticated Proxy Enabled: (0|1)/
  );

  if (match) {
    const [_, enabled, host, port, authenticated] = match;

    return {
      enabled: enabled === 'Yes',
      host,
      port,
      authenticated: authenticated === 'Yes',
    };
  } else {
    throw new Error(`Unexpected proxy config returned from networksetup`);
  }
}
