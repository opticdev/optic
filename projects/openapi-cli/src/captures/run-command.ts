import { ChildProcessWithoutNullStreams, spawn } from 'child_process';
import { createCommandFeedback } from '../commands/reporters/feedback';
import url from 'url';

export class RunCommand {
  constructor(
    private proxyUrl: string,
    private feedback: ReturnType<typeof createCommandFeedback>
  ) {}

  completed: boolean = false;
  process: ChildProcessWithoutNullStreams | undefined = undefined;

  async run(command: string): Promise<{ exitCode: number }> {
    const { port, hostname } = url.parse(this.proxyUrl);

    const [commandStart] = command.split(/ +/);

    const commandProcess = spawn(command, {
      cwd: process.cwd(),
      env: {
        ...process.env,
        http_proxy: `http://${hostname}:${port}`,
        https_proxy: `https://${hostname}:${port}`,
      },
      shell: true,
    });

    this.feedback.notable('Running command: ' + command);

    this.process = commandProcess;

    commandProcess.stdout.setEncoding('utf8');

    commandProcess.stdout.on('data', (data) => {
      const str = data.toString(),
        lines = str.split(/(\r?\n)/g);
      lines.forEach(
        (line) => line.trim() && this.feedback.logChild(commandStart, line)
      );
    });

    commandProcess.stderr.on('data', (data) => {
      const str = data.toString(),
        lines = str.split(/(\r?\n)/g);
      lines.forEach(
        (line) => line.trim() && this.feedback.logChild(commandStart, line)
      );
    });

    return new Promise((resolve) => {
      const endIt = (code: number | undefined) => {
        this.completed = true;
        resolve({ exitCode: commandProcess?.exitCode || 0 });
        commandProcess!.removeAllListeners('error');
        commandProcess!.removeAllListeners('data');
        commandProcess!.removeAllListeners('exit');
      };
      commandProcess!.once('exit', (code) => endIt(code || undefined));
      commandProcess!.once('end', endIt);
      commandProcess!.once('close', endIt);
    });
  }

  async kill() {
    if (!this.completed && this.process) {
      this.process.kill(1);
    }
  }
}
