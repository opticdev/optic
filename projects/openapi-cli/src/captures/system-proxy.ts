import url from 'url';
import { createCommandFeedback } from '../commands/reporters/feedback';
import { platform, runCommand } from '../shell-utils';

export class SystemProxy {
  constructor(
    private proxyUrl: string,
    private feedback: ReturnType<typeof createCommandFeedback>
  ) {}
  private stopCommand: (() => Promise<void>) | undefined = undefined;

  async start(device: string | undefined) {
    const { port } = url.parse(this.proxyUrl);

    if (platform === 'mac') {
      const name = await chooseInterfaceMac();

      await Promise.all([
        runCommand(`networksetup -setwebproxy "${name}" 127.0.0.1 ${port}`),
        runCommand(
          `networksetup -setsecurewebproxy "${name}" 127.0.0.1 ${port}`
        ),
        runCommand(
          `networksetup -setproxybypassdomains "${name}" "<-loopback>"`
        ),
      ]);

      this.feedback.notable(
        `Proxy running on ${this.proxyUrl}. System proxy updated`
      );

      this.stopCommand = async () => {
        await Promise.all([
          runCommand(`networksetup -setwebproxystate "${name}" off`),
          runCommand(`networksetup -setsecurewebproxystate "${name}" off`),
        ]);
        this.feedback.notable(`Mac System Proxy settings cleared`);
      };
    } else {
      this.feedback.notable(
        `Proxy running on ${this.proxyUrl}. System proxy updated`
      );
      console.log(
        `automatic proxy configuration is not supported on ${platform}`
      );
    }
  }
  async stop() {
    if (this.stopCommand) await this.stopCommand();
  }
}

export async function chooseInterfaceMac() {
  const network = await runCommand('scutil --nwi');
  const device = new RegExp(/Network interfaces: (en[0-9])/);

  const result = network.match(device);

  let devicePreference = result ? result[1] : 'en0';

  const hardware = await runCommand('networksetup -listallhardwareports');
  const regex = new RegExp(/(Hardware Port: (.*)\nDevice: (.*)\n)/);

  let name: string | undefined = undefined;
  hardware.split(/\n\s*\n/).forEach((entry) => {
    if (name) return;
    const match = regex.exec(entry);

    if (match && match[3] === devicePreference) {
      name = match[2];
    }
  });
  return name || 'Wi-Fi';
}
