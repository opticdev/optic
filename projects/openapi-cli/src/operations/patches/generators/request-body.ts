import { OperationDiffResult, OperationDiffResultKind } from '../../diffs';
import { OperationPatch, PatchOperationGroup, PatchImpact } from '..';
import { Operation } from '../..';
import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';

export function* requestBodyPatches(
  diff: OperationDiffResult,
  operation: Operation,
  context?: {}
): IterableIterator<OperationPatch> {
  if (diff.kind === OperationDiffResultKind.MissingRequestBody) {
    yield* missingRequestBodyPatches();
  } else if (diff.kind === OperationDiffResultKind.UnmatchedRequestBody) {
    yield* unmatchedRequestBodyPatches(diff, operation);
  }
}

function* missingRequestBodyPatches(): IterableIterator<OperationPatch> {
  const operationGroup = PatchOperationGroup.create(
    'make request body optional',
    {
      op: 'replace',
      path: jsonPointerHelpers.compile(['requestBody', 'required']),
      value: false,
    }
  );

  yield {
    description: 'make request body optional',
    impact: [PatchImpact.Addition, PatchImpact.BackwardsCompatible],
    groupedOperations: [operationGroup],
  };
}

function* unmatchedRequestBodyPatches(
  diff: OperationDiffResult & {
    kind: OperationDiffResultKind.UnmatchedRequestBody;
  },
  operation: Operation
): IterableIterator<OperationPatch> {
  const requestBody = operation.requestBody;

  if (!diff.contentType) return; // we can't add a content type if it's unknown
  const contentType = diff.contentType;

  let groupedOperations: PatchOperationGroup[] = [];
  if (!requestBody) {
    groupedOperations.push(
      PatchOperationGroup.create(`add request body to operation`, {
        op: 'add',
        path: jsonPointerHelpers.compile(['requestBody']),
        value: {
          content: {},
        },
      })
    );
  }

  groupedOperations.push(
    PatchOperationGroup.create(`add ${contentType} as content type`, {
      op: 'add',
      path: jsonPointerHelpers.compile(['requestBody', 'content', contentType]),
      value: {},
    })
  );

  yield {
    description: `add '${contentType}' body as a valid request body type`,
    impact: [PatchImpact.Addition, PatchImpact.BackwardsCompatible],
    groupedOperations,
  };
}
