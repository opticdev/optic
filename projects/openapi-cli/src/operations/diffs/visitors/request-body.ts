import { CapturedRequest } from '../../../captures';
import { OperationDiffResult, OperationDiffResultKind } from '../result';
import { OpenAPIV3 } from '../../../specs';

export function* visitRequestBody(
  request: CapturedRequest,
  spec?: OpenAPIV3.RequestBodyObject
): IterableIterator<OperationDiffResult> {
  if (!spec) {
    if (request.body) {
      // no request body was documented, but present in interaction
      yield {
        kind: OperationDiffResultKind.UnmatchedRequestBody,
        contentType: request.body.contentType,
      };
    }
    return; // no spec, nothing left to diff
  }

  if (!request.body) {
    if (spec.required) {
      yield {
        kind: OperationDiffResultKind.MissingRequestBody,
      };
    }

    return; // no request body, nothing left to diff
  }

  const bodySpec =
    request.body.contentType && spec.content[request.body.contentType];

  if (!bodySpec) {
    yield {
      kind: OperationDiffResultKind.UnmatchedRequestBody,
      contentType: request.body.contentType,
    };
  }
}
