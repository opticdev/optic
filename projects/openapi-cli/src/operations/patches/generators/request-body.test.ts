import { requestBodyPatches } from './request-body';
import {
  diffInteractionByOperation,
  OperationDiffResultKind,
} from '../../diffs';
import { OpenAPIV3 } from '../../../specs';
import { Operation, HttpMethods } from '../../';
import { CapturedInteraction, CapturedBody } from '../../../captures';
import { OperationPatch } from '..';

describe('requestBodyPatches', () => {
  it('generates a patch for an unspecified request body', () => {
    let operationWithoutRequestBody = operationFixture(null);
    let interactionWithRequestBody = interactionFixture(
      CapturedBody.fromJSON({}, 'application/json')
    );

    let [diff] = [
      ...diffInteractionByOperation(
        interactionWithRequestBody,
        operationWithoutRequestBody
      ),
    ];
    expect(diff.kind).toBe(OperationDiffResultKind.UnmatchedRequestBody);

    let patches = [...requestBodyPatches(diff, operationWithoutRequestBody)];
    expect(patches).toHaveLength(1);
    expect(patches).toMatchSnapshot();

    for (let patch of patches) {
      let patchedOperation = OperationPatch.applyTo(
        patch,
        operationWithoutRequestBody
      ).expect('operation patch should apply to operation');

      expect(patchedOperation).toMatchSnapshot();

      expect([
        ...diffInteractionByOperation(
          interactionWithRequestBody,
          patchedOperation
        ),
      ]).toHaveLength(0);
    }
  });

  it('generates a patch for a missing request body', () => {
    const operationWithRequestBody = operationFixture({
      content: {
        'application/json': {},
      },
      required: true,
    });

    const requestWithoutBody = interactionFixture(null);

    let [diff] = [
      ...diffInteractionByOperation(
        requestWithoutBody,
        operationWithRequestBody
      ),
    ];
    expect(diff.kind).toBe(OperationDiffResultKind.MissingRequestBody);

    let patches = [...requestBodyPatches(diff, operationWithRequestBody)];
    expect(patches).toHaveLength(1);
    expect(patches).toMatchSnapshot();

    for (let patch of patches) {
      let patchedOperation = OperationPatch.applyTo(
        patch,
        operationWithRequestBody
      ).expect('operation patch should apply to operation');

      expect(patchedOperation).toMatchSnapshot();

      expect([
        ...diffInteractionByOperation(requestWithoutBody, patchedOperation),
      ]).toHaveLength(0);
    }
  });

  it('generates a patch for a unmatched request body content type', () => {
    const operationWithJsonBody = operationFixture({
      content: {
        'application/json': {},
      },
      required: true,
    });

    const csvRequest = interactionFixture(
      CapturedBody.from('item1,item2', 'text/csv')
    );

    let [diff] = [
      ...diffInteractionByOperation(csvRequest, operationWithJsonBody),
    ];
    expect(diff.kind).toBe(OperationDiffResultKind.UnmatchedRequestBody);

    let patches = [...requestBodyPatches(diff, operationWithJsonBody)];
    expect(patches).toHaveLength(1);
    expect(patches).toMatchSnapshot();

    for (let patch of patches) {
      let patchedOperation = OperationPatch.applyTo(
        patch,
        operationWithJsonBody
      ).expect('operation patch should apply to operation');

      expect(patchedOperation).toMatchSnapshot();
      expect([
        ...diffInteractionByOperation(csvRequest, patchedOperation),
      ]).toHaveLength(0);
    }
  });

  it('does not generate new patches for captured bodies with an unknown content type', () => {
    const operationWithJsonBody = operationFixture({
      content: {
        'application/json': {},
      },
      required: true,
    });

    const missingContentTypeRequest = interactionFixture(
      CapturedBody.from('test-body')
    );

    let [diff] = [
      ...diffInteractionByOperation(
        missingContentTypeRequest,
        operationWithJsonBody
      ),
    ];
    expect(diff.kind).toBe(OperationDiffResultKind.UnmatchedRequestBody);

    let patches = [...requestBodyPatches(diff, operationWithJsonBody)];
    expect(patches).toHaveLength(0);
  });

  it('does not generate new patches in the context of a non 2xx or 3xx status code', () => {
    const operationWithoutRequestBody = operationFixture(null);

    const request = interactionFixture(
      CapturedBody.fromJSON({ id: 'test' }, 'application/json')
    );

    let [diff] = [
      ...diffInteractionByOperation(request, operationWithoutRequestBody),
    ];
    expect(diff.kind).toBe(OperationDiffResultKind.UnmatchedRequestBody);

    let patches = [
      ...requestBodyPatches(diff, operationWithoutRequestBody, {
        statusCode: '101',
      }),
    ];
    expect(patches).toHaveLength(0);

    patches = [
      ...requestBodyPatches(diff, operationWithoutRequestBody, {
        statusCode: '200',
      }),
    ];
    expect(patches).toHaveLength(1);

    patches = [
      ...requestBodyPatches(diff, operationWithoutRequestBody, {
        statusCode: '302',
      }),
    ];
    expect(patches).toHaveLength(1);

    patches = [
      ...requestBodyPatches(diff, operationWithoutRequestBody, {
        statusCode: '400',
      }),
    ];
    expect(patches).toHaveLength(0);
  });
});

function operationFixture(
  requestBody: OpenAPIV3.RequestBodyObject | null
): Operation {
  return {
    requestBody,
    method: HttpMethods.POST,
    pathPattern: '/some-path',
    responses: {
      '200': {
        description: 'success',
      },
    },
  };
}

function interactionFixture(
  requestBody: CapturedBody | null
): CapturedInteraction {
  return {
    request: {
      host: 'optic.test',
      method: HttpMethods.POST,
      path: '/some-path',
      body: requestBody,
    },
    response: {
      statusCode: '200',
      body: null,
    },
  };
}
