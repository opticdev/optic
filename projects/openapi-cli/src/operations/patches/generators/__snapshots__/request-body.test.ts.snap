// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requestBodyPatches generates a patch for a missing request body 1`] = `
Array [
  Object {
    "description": "make request body optional",
    "groupedOperations": Array [
      Object {
        "intent": "make request body optional",
        "operations": Array [
          Object {
            "op": "replace",
            "path": "/requestBody/required",
            "value": false,
          },
        ],
      },
    ],
    "impact": Array [
      "Addition",
      "BackwardsCompatible",
    ],
  },
]
`;

exports[`requestBodyPatches generates a patch for a missing request body 2`] = `
Object {
  "method": "post",
  "pathPattern": "/some-path",
  "requestBody": Object {
    "content": Object {
      "application/json": Object {},
    },
    "required": false,
  },
  "responses": Object {
    "200": Object {
      "description": "success",
    },
  },
}
`;

exports[`requestBodyPatches generates a patch for a unmatched request body content type 1`] = `
Array [
  Object {
    "description": "add 'text/csv' body as a valid request body type",
    "groupedOperations": Array [
      Object {
        "intent": "add text/csv as content type",
        "operations": Array [
          Object {
            "op": "add",
            "path": "/requestBody/content/text~1csv",
            "value": Object {},
          },
        ],
      },
    ],
    "impact": Array [
      "Addition",
      "BackwardsCompatible",
    ],
  },
]
`;

exports[`requestBodyPatches generates a patch for a unmatched request body content type 2`] = `
Object {
  "method": "post",
  "pathPattern": "/some-path",
  "requestBody": Object {
    "content": Object {
      "application/json": Object {},
      "text/csv": Object {},
    },
    "required": true,
  },
  "responses": Object {
    "200": Object {
      "description": "success",
    },
  },
}
`;

exports[`requestBodyPatches generates a patch for an unspecified request body 1`] = `
Array [
  Object {
    "description": "add 'application/json' body as a valid request body type",
    "groupedOperations": Array [
      Object {
        "intent": "add request body to operation",
        "operations": Array [
          Object {
            "op": "add",
            "path": "/requestBody",
            "value": Object {
              "content": Object {},
            },
          },
        ],
      },
      Object {
        "intent": "add application/json as content type",
        "operations": Array [
          Object {
            "op": "add",
            "path": "/requestBody/content/application~1json",
            "value": Object {},
          },
        ],
      },
    ],
    "impact": Array [
      "Addition",
      "BackwardsCompatible",
    ],
  },
]
`;

exports[`requestBodyPatches generates a patch for an unspecified request body 2`] = `
Object {
  "method": "post",
  "pathPattern": "/some-path",
  "requestBody": Object {
    "content": Object {
      "application/json": Object {},
    },
  },
  "responses": Object {
    "200": Object {
      "description": "success",
    },
  },
}
`;
